### DISCUSSION TITLE
What prevents ghc@8.8 from using a bottle (with a non-default prefix)?

### DISCUSSION THREAD

**Discussion by kthchew** [Homebrew/discussions#888] [Linux]
I have the non-default Homebrew prefix, `~/.brew`. I understand that with this configuration, some formulae that are not relocatable will need to be built from source.

One such formula is `ghc@8.8`. When installing it without any additional flags, I get this message:

```
❯ brew install ghc@8.8
Warning: Building ghc@8.8 from source as the bottle needs:
- HOMEBREW_CELLAR: /usr/local/Cellar (yours is /Users/kenneth/.brew/Cellar)
- HOMEBREW_PREFIX: /usr/local (yours is /Users/kenneth/.brew)
```

I was wondering whether this was something that I could try to resolve, so first I used `brew install --build-bottle ghc@8.8`, which (after more than 2 hours) was successful. Then, I used `brew bottle --verbose ghc@8.8` to see if I could fix any errors I saw. This was the output I got:

```
❯ brew bottle --verbose ghc@8.8
git config --replace-all homebrew.devcmdrun true
==> Determining ghc@8.8 bottle rebuild...
==> Bottling ghc@8.8--8.8.4_1.big_sur.bottle.1.tar.gz...
tar cf /Users/kenneth/.brew/Library/Taps/homebrew/homebrew-cask-versions/ghc@8.8--8.8.4_1.big_sur.bottle.1.tar ghc@8.8/8.8.4_1
gzip -f ghc@8.8-bottle.tar
==> Detecting if ghc@8.8--8.8.4_1.big_sur.bottle.1.tar.gz is relocatable...
./ghc@8.8--8.8.4_1.big_sur.bottle.1.tar.gz
  bottle do
    rebuild 1
    sha256 cellar: :any, big_sur: "c758212d48edb54296c40066ae545b15d988443f4c211061f0321160c91f55f9"
  end
```

Strangely enough, I don't see any errors; in fact, it looks like it decided that it _was_, in fact, relocatable (am I interpreting this output correctly?). 

I'm not here to ask for it to be fixed, but I was wondering why Homebrew decided to build `ghc@8.8` from source. If `ghc@8.8` is, in fact, relocatable, then what other criteria decides whether Homebrew builds from source or tries to use a bottle when a non-default prefix is used?

Note: I'm asking specifically about `ghc@8.8` because it's what I tested; `ghc` also decides to build from source, and I assume it would be the same, but I haven't tried it myself due to the long build times.

**Comment by jonchang** [comment]
Try running `brew bottle --verbose --debug FORMULA`. The logs for building ghc@8.8 have unfortunately expired so we can't find the original reasoning why it's not relocatable. We only recently permitted formulae to have different relocatable status for macOS versions, so it's possible that it's relocatable on e.g. Big Sur but not on Catalina.

**Reply by kthchew** [reply]
I tried that, and the logs are at https://gist.github.com/kthchew/deb962a3d3b0fb0b5aebb188392a5dc2. I'm not completely sure, but I don't think I see anything of interest there.

> We only recently permitted formulae to have different relocatable status for macOS versions, so it's possible that it's relocatable on e.g. Big Sur but not on Catalina.

Ah, makes sense, thank you. I'm on Big Sur. If this is indeed the case, would this be automatically resolved the next time the formula updates?

**Reply by danielnachun** [reply]
That log indicates the bottle is relocatable, and my understanding of the way `ghc` is works is that it should always be relocatable.  `ghc` takes a very long time to build, and it's possible that this has something to do with why it was not given the proper cellar.  It looks like all versions of `ghc` lost their relocatable status, so there may need to be more investigation here.

**Reply by kthchew** [reply]
> It looks like all versions of `ghc` lost their relocatable status, so there may need to be more investigation here.

In this case, should I make an issue to track this outside of just this discussion?
